explainer.md

```diff
@@ -215,6 +243,162 @@ third-party then signs either the exchanges in a package or the package as a
+requests that aren't under the browser's control and therefore use a separate,
+probably ephemeral, cookie jar, cache, and set of service workers.
+
+Each signature with a valid certificate chain is passed on to the next step.
+
+### Signature verification
+
+Once the certificates are validated and enough of the
+`application/signed-exchange` resource is received to parse the claimed
+signed-exchange headers, the client extracts the logical URL from those headers
+and then tries to find a valid signature over the headers that is trusted for
+the claimed origin. If none of the signatures are valid, it either
+
+1. redirects to the logical URL as if the whole signed exchange were a 302
+   response, or
+2. fails with a network error.
+
+We're not yet certain which behavior is best. The first is slightly more
+resilient to clock skew, while the second encourages intermediates to get their
+implementations right.
+
+If the client does find a valid signature, it "releases" the headers for further
+processing and starts validating
+[mi-sha256](https://tools.ietf.org/html/draft-thomson-http-mice-02#section-2)
+records into a response stream as they arrive.
+
+### Prefetching stops here
+
+At this point, the client's behavior depends on whether the signed exchange was
+requested as a [prefetch](https://w3c.github.io/resource-hints/#prefetch). To
+satisfy the [privacy-preserving prefetch](#privacy-preserving-prefetch) use
+case, prefetches can't fully load the logical URL, which would create an HTTP
+cache entry that would be visible to the logical URL's server.
+
+Prefetches can and should process any `Link: <>; rel=preload` headers they find,
+as prefetches. If those point at signed exchanges, this process repeats.
+
+### Caching the signed response
+
+If the signed exchange was requested as a navigation or subresource (i.e. *not*
+prefetches), the client tries to cache the request→response pair it contains.
+
+This *doesn't* happen if:
+
+* The signed exchange's request headers aren't sufficiently similar (TBD) to the
+  request headers the client would use for a normal request in the same context.
+  This prevents a malicious intermediate from sticking the wrong
+  content-negotiated resource in the HTTP cache.
+* There's a response in the HTTP (or preload?) cache with a newer Date header
+  than the signed exchange's response. This prevents some downgrade attacks.
+
+In either of these cases, the client just skips to the redirect in the next
+step.
+
+If we're still here, the signed exchange is put into the [preload
+cache](https://github.com/whatwg/fetch/issues/590) and, if the response headers
+allow it (and review of this explainer indicates it's a good idea), the [HTTP
+cache](https://tools.ietf.org/html/rfc7234).
+
+If we put the signed exchange in the HTTP cache, its freshness has to be bounded
+by the shorter of the normal HTTP cache lifetime or the signature's expiration.
```

Comment:
Does this imply that the signature's expiration has somewhat sticky effect on the resources that are installed from the signed exchange / bundle?  For some use cases like offline PWA installation this may not make a lot sense, or may have some inconsistency?  E.g. I think we'd want to keep the PWA's SW and assets installed from a bundle/exchange valid even after the signature expires, is that right?

Reply:
I see reg: offline use case.  What do you think about the following two cases:

* UA populates HTTP cache with resources extracted from the signed exchange, and they expire when the signature of them expire
* SW populates Cache Storage with resources come from the signed exchange (and SW may know the fact or not), and they will be around even after the signature of the original exchange expires

My mental model has been that the expiration matters when we process the signed exchange, but once if it's processed the resources that come from the signed exchange could just look similar to others. I think I'm fine with the current text, but wanted to note that (partly because not sticking the signature info is easier to implement).